com-api/src/lib.rs

```diff
@@ -0,0 +1,179 @@
+//! # API Design principles
+//!
+//! - We stick to the builder pattern down to a single service
+//! - We make all elements mockable. This means we provide traits for the building blocks.
+//!   We strive for enabling trait objects for mockable entities.
+//! - We allow for the usage of heap memory during initialization phase (offer, connect, ...)
+//!   but prevent heap memory usage during the running phase. Any allocations during the run phase
+//!   must happen on preallocated memory chunks.
+//! - We support data provision on potentially uninitialized memory for efficiency reasons.
+//! - We want to be type safe by deriving as much type information as possible from the interface
+//!   description.
+//! - We want to design interfaces such that they're hard to misuse.
+//! - Data communicated over IPC need to be position independent.
+//! - Data may not reference other data outside its provinence
+//! - We provide simple defaults for customization points
+//! - Sending / receiving / calling is always done explicitly.
+//! - COM API does not enforce the necessity for internal threads or executors.
+//!
+//! # Lower layer implementation principles
+//!
+//! - We add safety nets to detect ABI incompatibilities
+//!
+//! # Supported IPC ABI
+//!
+//! - Primitives
+//! - Static lists / strings
+//! - Dynamic lists / strings
+//! - Key-value
+//! - Structures
+//! - Tuples
+
+use std::fmt::Debug;
+use std::ops::{Deref, DerefMut};
+use std::path::Path;
+
+#[derive(Debug)]
+pub enum Error {
+    Fail,
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+pub trait Builder {
+    type Output;
+    /// Open point: Should this be &mut self so that this can be turned into a trait object?
+    fn build(self) -> Self::Output;
+}
+
+pub trait Runtime {}
+
+pub trait RuntimeBuilder: Builder
+where
+    <Self as Builder>::Output: Runtime,
+{
+    fn load_config(&mut self, config: &Path) -> &mut Self;
+}
+
+/// Determines whether a type may be trivially moved without invalidating inner integrity
+pub trait Reloc {}
+
+impl Reloc for () {}
+impl Reloc for u32 {}
```

Comment:
We might want to make this an `unsafe` trait.

> We can use unsafe to implement an unsafe trait. A trait is unsafe when at least one of its methods **has some invariant that the compiler can’t verify**.[...] - [Rust Book: Implementing an Unsafe Trait](https://doc.rust-lang.org/book/ch20-01-unsafe-rust.html)


```suggestion
/// Determines whether a type may be trivially moved without invalidating inner integrity
///
/// # Safety
///
/// The implementor has to ensure that types implementing [Reloc]
/// stay valid after being moved to a new address.
pub unsafe trait Reloc {}

unsafe impl Reloc for () {}
unsafe impl Reloc for u32 {}
```

Reply:
That might actually be necessary as the trait promises (without the compiler being able to check) that the type is movable without invalidating its invariants.

I was reiterating on the `Unpin` trait, and I have to say that I'm still not convinced that we can't just reuse that. It would be great since it's auto-implemented by the compiler based on whether the parts of the struct are `Unpin`. And it's usage is in fact exactly what we want: We need to be able to distinguish between movable types and those who can't be moved to be able to support data exchange between processes.

One could now argue that using shared memory isn't moving technically, but since the shared memory block will most likely not be mapped to the same virtual address space, the resulting effect is basically the same as moving data of that type by means of a bitwise copy.

I therefore would want to get further input on the semantic difference between `Unpin` and `Reloc` and this should then also be thoroughly documented.

Reply:
Hm I agree that it would be great to reuse something which can be auto-inferred by the compiler. Though I would want to think a bit more about the implications:

* A `Reloc` type can always be moved freely, whether it was used or not.
* A `Unpin` type can always be moved freely, whether someone accessed it through a `Pin` previously or not.
* A `!Unpin` type can be moved freely before it is ever accessed through a `Pin`. Once the first `Pin` on it was created, it no longer can be moved (imagine a Autrian Autobahn vignette which can be passed around until it gets glued for the first time to a windshield).

From this comparison, I guess we would want to require `Unpin`. The questions are:

1. Can we think of any type which is `Unpin` but does not fulfill our idea of `Reloc`?
2. Can we think of any type which is `!Unpin` but fulfills everything which we have in mind for `Reloc`?

If we can reason "No" for both questions, then we could argue to recycle `Unpin`.